### PR TITLE
8319374: JFR: Remove instrumentation for exception events

### PR DESCRIPTION
**Pull request by egahlin**
Could I have a review of a PR that removes the bytecode instrumentation for the exception events.

Testing: jdk/jdk/jfr + tier1 + tier2

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319374](https://bugs.openjdk.org/browse/JDK-8319374): JFR: Remove instrumentation for exception events (**Enhancement** - P3)

### Reviewers
 * [Markus Grönlund](https://openjdk.org/census#mgronlun) (@mgronlun - **Reviewer**)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16493/head:pull/16493` \
`$ git checkout pull/16493`

Update a local copy of the PR: \
`$ git checkout pull/16493` \
`$ git pull https://git.openjdk.org/jdk.git pull/16493/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16493`

View PR using the GUI difftool: \
`$ git pr show -t 16493`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16493.diff">https://git.openjdk.org/jdk/pull/16493.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16493#issuecomment-1795102338)

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/event/ThrowableTracer.java [hunk 1]
```diff
@@ -22,47 +22,50 @@
...
-    public static void traceError(Error e, String message) {
-        if (e instanceof OutOfMemoryError) {
+    public static void enable() throws NoSuchFieldException, SecurityException, IllegalArgumentException, IllegalAccessException {
+        Field field = Throwable.class.getDeclaredField("tracing");
+        field.setAccessible(true);
+        field.setBoolean(null, true);
+    }
+
+    public static void traceError(Class<?> clazz, String message) {
+        if (OutOfMemoryError.class.isAssignableFrom(clazz)) {
```
Comment (AlanBateman) [inline src/java.base/share/classes/jdk/internal/event/ThrowableTracer.java:44]: StackOverflowError is likely problematic too, maybe it should be VirtualMachineError.

File: src/java.base/share/classes/jdk/internal/event/ThrowableTracer.java [hunk 2]
```diff
@@ -22,47 +22,50 @@
...
+/**
+ * Helper class for exception events.
+ */
 public final class ThrowableTracer {
 
     private static final AtomicLong numThrowables = new AtomicLong();
 
-    public static void traceError(Error e, String message) {
-        if (e instanceof OutOfMemoryError) {
+    public static void enable() throws NoSuchFieldException, SecurityException, IllegalArgumentException, IllegalAccessException {
```
Comment (AlanBateman) [inline src/java.base/share/classes/jdk/internal/event/ThrowableTracer.java:37]: SecurityException and IllegaArgumentException are unchecked so don't need them in the throws declaration.

File: src/java.base/share/classes/jdk/internal/event/ThrowableTracer.java (hunk 1 above)
Comment (egahlin) [reply src/java.base/share/classes/jdk/internal/event/ThrowableTracer.java:44]: I remember asking the same question ten years ago, when Nils did the original implementation, but I think it was only needed for OOM, because it creates an infinite loop when the event object was allocated, which resulted in a StackOverflowError instead OOM.

Some OOM tests failed with JFR enabled.

The event object allocation has been removed, but I think we can run into the allocation recursion by other means. I looked into it a few years ago, but I don't remember exactly why it failed.

If a SOE happens, I think we are fine. There is something that prevents infinite recursion when the SOE object is created. Perhaps it is preallocated? I prefer to not change the behavior, at least not in this PR.

File: src/java.base/share/classes/jdk/internal/event/ThrowableTracer.java (hunk 2 above)
Comment (egahlin) [reply src/java.base/share/classes/jdk/internal/event/ThrowableTracer.java:37]: Will fix.

File: src/java.base/share/classes/jdk/internal/event/ThrowableTracer.java (hunk 1 above)
Comment (egahlin) [reply src/java.base/share/classes/jdk/internal/event/ThrowableTracer.java:44]: I filed an issue to investigate if there is problem with SOE, or if the OOM check is really needed now.
https://bugs.openjdk.org/browse/JDK-8319579

File: src/java.base/share/classes/java/lang/Throwable.java [hunk 3]
```diff
@@ -118,6 +119,11 @@ public class Throwable implements Serializable {
     @java.io.Serial
     private static final long serialVersionUID = -3042686055658047285L;
 
+    /**
+     * Flag that determines if exceptions should be traced by JFR
+     */
+    static volatile boolean tracing;
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/lang/Throwable.java:125]: The use-sites in Error check Throwable.tracing so not  immediately obvious that it's JFR tracing. Maybe it should be renamed to jfrTracing or an accessor of that name added so it's a bit clearer at the use-sites.

File: src/java.base/share/classes/jdk/internal/event/ThrowableTracer.java (hunk 1 above)
Comment (AlanBateman) [reply src/java.base/share/classes/jdk/internal/event/ThrowableTracer.java:44]: > I filed an issue to investigate if there is a problem with SOE, or if the OOM check is really needed now. https://bugs.openjdk.org/browse/JDK-8319579
> 
> Regardless of outcome, It would be good to document the results of the investigation in the code.

Okay, the scenario I was wondering about is where traceError with SOE and whether that could trigger a loop too.

Comment (egahlin) [reply src/java.base/share/classes/jdk/internal/event/ThrowableTracer.java:44]: I understand.

File: src/java.base/share/classes/java/lang/Throwable.java (hunk 3 above)
Comment (egahlin) [reply src/java.base/share/classes/java/lang/Throwable.java:125]: Will change

**Review by mgronlun — APPROVED** [review]

**Review by AlanBateman — APPROVED** [review]
The changes looks okay and will track the follow-on issue to look at SOE.

File: src/java.base/share/classes/jdk/internal/event/ThrowableTracer.java [hunk 4]
```diff
@@ -22,47 +22,50 @@
...
 
-    public static void traceThrowable(Throwable t, String message) {
-        EventConfiguration eventConfiguration = EventConfigurations.EXCEPTION_THROWN;
-        if (eventConfiguration.isEnabled()) {
-            long timestamp = EventConfiguration.timestamp();
-            ExceptionThrownEvent.commit(timestamp, message, t.getClass());
+    public static void traceThrowable(Class<?> clazz, String message) {
+        if (ExceptionThrownEvent.enabled()) {
+            long timestamp = ExceptionThrownEvent.timestamp();
+            ExceptionThrownEvent.commit(timestamp, message, clazz);
```
Comment (dholmes-ora) [inline src/java.base/share/classes/jdk/internal/event/ThrowableTracer.java:62]: Just a drive-by observation, but now you get the timestamp from the event it seems strange to pull it out and then pass it back to commit, instead of having a version of commit that automatically uses the internal timestamp.

Comment (egahlin) [reply src/java.base/share/classes/jdk/internal/event/ThrowableTracer.java:62]: I agree, and I have looked into it, but I think it's better to do that refactorization seperately.

Comment (dfuch) [reply src/java.base/share/classes/jdk/internal/event/ThrowableTracer.java:62]: Just for my own understanding: in this particular case the time stamp is meaningless because the duration is expected to be 0 (or close to it) since nothing happens between the time the timestamp is taken and the time the event is committed. Is that correct?

Comment (egahlin) [reply src/java.base/share/classes/jdk/internal/event/ThrowableTracer.java:62]: Yes, the event has no duration.

Before https://bugs.openjdk.org/browse/JDK-8239508 we wrote timestamp and duration for all Java events. For exception events, we wrote duration = 0. Now we can differentiate between durational and instant events, which means bytecode can be generated to take the timestamp in the commit method.
